crypto/hpke/hpke.c

```diff
@@ -0,0 +1,2579 @@
+            if (HPKE_MSMATCH(st, "0x1")) { kdf = 1; }
+            if (HPKE_MSMATCH(st, "0x2")) { kdf = 2; }
+            if (HPKE_MSMATCH(st, "0x3")) { kdf = 3; }
+            if (HPKE_MSMATCH(st, "1")) { kdf = 1; }
+            if (HPKE_MSMATCH(st, "2")) { kdf = 2; }
+            if (HPKE_MSMATCH(st, "3")) { kdf = 3; }
+        } else if (kem != 0 && kdf != 0 && aead == 0) {
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_AES128GCM)) { aead = 1; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_AES256GCM)) { aead = 2; }
+            if (HPKE_MSMATCH(st, OSSL_HPKE_AEADSTR_CP)) { aead = 3; }
+            if (HPKE_MSMATCH(st, "0x01")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "0x02")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "0x03")) { aead = 3; }
+            if (HPKE_MSMATCH(st, "0x1")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "0x2")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "0x3")) { aead = 3; }
+            if (HPKE_MSMATCH(st, "1")) { aead = 1; }
+            if (HPKE_MSMATCH(st, "2")) { aead = 2; }
+            if (HPKE_MSMATCH(st, "3")) { aead = 3; }
+        }
+        st = strtok(NULL, ",");
+    }
+    OPENSSL_free(instrcp);
+    if ((st != NULL && labels > 3) || kem == 0 || kdf == 0 || aead == 0) {
+        return 0;
+    }
+    suite->kem_id = kem;
+    suite->kdf_id = kdf;
+    suite->aead_id = aead;
+    return 1;
+}
+
+/*
+ * @brief tell the caller how big the cipertext will be
+ *
+ * AEAD algorithms add a tag for data authentication.
+ * Those are almost always, but not always, 16 octets
+ * long, and who knows what'll be true in the future.
+ * So this function allows a caller to find out how
+ * much data expansion they'll see with a given suite.
+ *
+ * @param suite is the suite to be used
+ * @param enclen points to what'll be enc length
+ * @param clearlen is the length of plaintext
+ * @param cipherlen points to what'll be ciphertext length
+ * @return 1 for success, otherwise failure
+ */
+static int hpke_expansion(OSSL_HPKE_SUITE suite,
+                          size_t *enclen,
+                          size_t clearlen,
+                          size_t *cipherlen)
+{
+    uint16_t aead_ind = 0;
+    uint16_t kem_ind = 0;
+
+    if (cipherlen == NULL || enclen == NULL) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    if (hpke_suite_check(suite) != 1) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    aead_ind = aead_iana2index(suite.aead_id);
+    if (aead_ind == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *cipherlen = clearlen + hpke_aead_tab[aead_ind].taglen;
+    kem_ind = kem_iana2index(suite.kem_id);
+    if (kem_ind == 0) {
+        ERR_raise(ERR_LIB_CRYPTO, ERR_R_INTERNAL_ERROR);
+        return 0;
+    }
+    *enclen = hpke_kem_tab[kem_ind].Nenc;
+    return 1;
+}
+
+static int hpke_seq2buf(uint64_t seq, unsigned char *buf, size_t blen)
```

Comment:
Why not just do it this way? (OpenSSL doesnt use htonl anywhere else except one place where it is conditionally used)..

````
    memcpy(out, aead->base_nonce, aead->base_noncelen);
    for (i = outlen - 1; i != 0; --i) {
        out[i] ^= (seq & 0xFF);
        seq >>= 8;
    }
````

Reply:
did more or less that in the end